### ISSUE TITLE
Jobs not being submitted to sites with RSEs ending with _Disk

### ISSUE THREAD

**Issue by hassan11196** [dmwm/WMCore#12012]
**Impact of the bug**
WMAgent

**Describe the bug**
Jobs for a workflow that reads Pileup locally i.e. `TrustPUSitelists: false` will not be submitted to the site if its  `phedex_name` name ends with `_Disk` Because the [possibleSites](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/DataStructs/WorkQueueElement.py#L37) function used by `WorkQueue` in the agent does not convert the `phedex_name`  to `psn_name` as defined in CRIC[1]. 

The [possibleSites](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/DataStructs/WorkQueueElement.py#L37)  method checks the pileup locations and the `NoPileupUpdate` (which I am assuming is linked to `TrustPUSitelists` from WF Config). and returns the intersection between `SiteWhitelist` and the Pileup data locations. The Pileup data locations are not converted from the Storage Element to Computing Element leading to this issue.

## How to reproduce it
For Example:
`cmsunified_task_JME-Run3Winter24wmLHEGS-00018__v1_T_240524_171256_5046`

This workflow has very high priority (310001) but only runs at CERN. 
![image](https://github.com/dmwm/WMCore/assets/33365802/d9de7a8a-9b07-4884-ad08-8dda5f442be6)
No Jobs submitted to FNAL.

Jobs are not submitted to FNAL even though it has lower priority wfs running because of the [possibleSites](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/DataStructs/WorkQueueElement.py#L37) method.

Example of One WorkQueue picked up by an FNAL Agent, the intersection of `SiteWhitelist` and `PileupData` will only return `T2_CH_CERN`.
```
        "SiteWhitelist": [
            "T1_US_FNAL",
            "T2_CH_CERN",
            "T2_CH_CERN_HLT",
            "T2_CH_CERN_P5"
        ],
        "SiteBlacklist": [],
        "StartPolicy": "MonteCarlo",
        "EndPolicy": {
            "policyName": "SingleShot"
        },
        "Priority": 310001,
        "PileupData": {
            "/MinBias_TuneCP5_13p6TeV-pythia8/Run3Winter24GS-133X_mcRun3_2024_realistic_v7-v1/GEN-SIM": [
                "T2_CH_CERN",
                "T1_US_FNAL_Disk"
            ]
        },
```

## Background

This problem has become apparent now because previously for minBias/heavy read pileups that do not read data remotely were only assigned to a single site. A bug in unified, caused workflows to be only assigned to a single site. We usually keep these pileups at CERN and FNAL, this bug caused the workflow to only be assigned to CERN. I fixed this bug and workflows started being assigned to both CERN and FNAL. [link](https://gitlab.cern.ch/CMSProductionReprocessing/WmAgentScripts/-/merge_requests/1418).

Previously since the wf was only being assigned to CERN, it was running at CERN so we did not notice this, but now the Workflows are assigned to both sites but only running at CERN.

## Verification

To verify this that no jobs whose wf configuration had `TrustPUSitelists: false` and pileup at `T1_US_FNAL_Disk` were running. I Compiled a list of all workflows that run at FNAL and CERN for the previous year. (2023-06-06 to 2024-06-06). 

There were a total of **1866** wfs with jobs whose `data.Site=T1_US_FNAL` and `data.Type=Production`.
**1664** had Pileup Datasets mentioned in the Steps/Tasks
**0** wfs had Pileup Datasets mentioned in the Steps/Tasks and `TrustPUSitelists`:`False`
![image](https://github.com/dmwm/WMCore/assets/33365802/fe465c00-10e6-47f4-ba8f-38872ac09932)

There were a total of **1600** wfs with jobs whose `data.Site=T2_CH_CERN`  and `data.Type=Production`.
**620** had Pileup Datasets mentioned in the Steps/Tasks
**212** wfs had Pileup Datasets mentioned in the Steps/Tasks and `TrustPUSitelists`:`False`
![image](https://github.com/dmwm/WMCore/assets/33365802/6e9d782a-364c-4beb-86c8-19419fa82a7e)

I assume the same issue effects workflows that have input Data and reading locally, i,e `TrustSitelists: False`, but I have not verified it yet.

- Assign a workflow with `TrustPUSitelists:false` only to a site whose storage element ends with `_Disk`.
- The Workflow will remain stuck in Acquired state and agents will not create jobs for it.

**Expected behavior**
Jobs to be properly submitted to all sites in the site whitelist even if the SE and CE names are different i.e "T1_US_FNAL_Disk",
"T1_US_FNAL".  
I can clean up the notebook I used to calculate the workflows. I was able to get data from Monit thanks to @nikodemas 

FYI @amaltaro @anpicci @haozturk

[1]https://cms-cric.cern.ch/api/cms/site/query/?json&preset=data-processing

**Comment by anpicci** [comment]
@amaltaro do I remember correctly that there was a reason to not consider site ending with `_Disk` as good ones? Or am I confusing things?

**Comment by amaltaro** [comment]
@hassan11196 having a careful look at this issue, I agree that the `possibleSites()` function returns inconsistent information, where `SiteWhitelist` and `SiteBlacklist` are actually a list of sites. But `Inputs` and `PileupData` actually contain a list of locations (RSEs).

Now looking for places that use this `possibleSites()` method, we can see (which does the PSN2PNN conversion of the output):
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/WorkQueue.py#L454-L455

but this module does not convert it:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/WorkQueueBackend.py#L385

In addition, I find apparently inconsistent use as well in this module:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/Policy/Start/StartPolicyInterface.py#L254
as we can see for `getDatasetLocations()`, `getDatasetLocationsFromMSPileup()` and `blockLocationRucioPhedex()`.

Lastly, we need to make sure that Sites (PSNs) and Storage (RSEs/PNNs/SEs) are properly considered in this CouchDB list function:
https://github.com/dmwm/WMCore/blob/master/src/couchapps/WorkQueue/lists/workRestrictions.js

Having said it all, I feel like we need someone to actually review everything that I said, review the code, test whatever is needed, such that we can find out where we have problems and whether objects should correspond to storage location (RSE/PNN) or CPU location (site/PSN). Please let me know if you think I am missing anything.

**Comment by hassan11196** [comment]
Thank you for the investigation, Alan. How do you recommend proceeding with the next steps?
 From your reply, I understand that we have the following 3 action items:
1. Identify all places where Sites / RSEs are used.
2. Decide whether it should be Site/PSN or RSE/PNN.
3. Implement fix (fixes?) that are consistent on all levels we use sites/PSN.

**Comment by amaltaro** [comment]
As we discussed yesterday in the WMCore meeting, we are going to get first the pileup issue under control, which we have already properly identified. In a later stage/ticket, we can revisit site vs location (PSN vs PNN) across WMCore and evaluate if there are other places in need of bug-fixing as well.


I wanted to have a full example of WQE configuration, which I found on https://cmsweb.cern.ch/couchdb/workqueue/_design/WorkQueue/_rewrite/element/29a66613a07a655cbda0bf0787633487

As one can see, this is how it is currently using sites vs locations:
```
    "Inputs": {
      "/SinglePhoton_Pt-200To500-gun/Run3Winter24GS-133X_mcRun3_2024_realistic_v7-v2/GEN-SIM#8fa9810e-0de6-42ce-85fc-183e6becae97": [
        "T1_US_FNAL",
        "T3_US_FNALLPC"
      ]
    },
    "NoInputUpdate": false,
    "NoPileupUpdate": false,
    "SiteWhitelist": [
      "T1_US_FNAL",
      "T2_CH_CERN",
      "T2_CH_CERN_HLT",
      "T2_CH_CERN_P5"
    ],
    "PileupData": {
      "/MinBias_TuneCP5_13p6TeV-pythia8/Run3Winter24GS-133X_mcRun3_2024_realistic_v7-v1/GEN-SIM": [
        "T2_CH_CERN",
        "T1_US_FNAL_Disk"
      ]
    },
```

in summary:
* this element does not have primary or secondary AAA enabled
* input data location is actually a list of PSNs
* site list is also a list of PSNs
* but pileup data contains a list of locations (PNNs/RSEs)

**Comment by khurtado** [comment]
Reopen issue, based on discussion on MM:
https://mattermost.web.cern.ch/cms-o-and-c/pl/ng3zisdcrig3igobk6qkw4txdr
